aws-sdk-s3.go

```diff
@@ -0,0 +1,46 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+func GetS3Client() *s3.S3 {
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}))
+
+	svc := s3.New(sess)
+
+	return svc
+}
+
+// Created soely for the purpose of testing
+// During the test, I need to mock the object passed into AwsSdkCountObjectsInBucket
+// If it is *s3.S3, all functions would need to be mocket
+// With a defined interface, only the method used needs to be mocked
+// As long as the object of type *s3.S3 is passed implements ListObjectsV2, this works
```

Comment:
Some notes for me to remember this when I have to do it in real life.

Reply:
Is there a Go equivalent to mocking the import and then providing return values when the import's public functions are used? Similar to how one would it in python:

```python
import boto3

@patch('somepath.boto3')
def test_my_func(mock_boto3):
    # Arrange
    mock_boto3.client.S3.list_objects_v2.return_value = {'Name':'my_object'}
    # Act
    result = my_func(mock_boto3)
    # Assert
    assert 1 == result
````

Go seems to be:
1. Make a struct
2. Define necessary functions so that the struct can satisfy the interface that is param to the function (where the functions return the mocked values I want)
3. Create an object of type struct and pass it in